### DISCUSSION TITLE
Regarding C++ language standard

### DISCUSSION THREAD

**Discussion by stalkyr** [microsoft/vcpkg#26226] [Ideas]
I want to build libraries with C++17 API. 

Some library, `mongo-cxx-driver` for example, can build with C++11 with polyfills or build with C++17 directly without depending on others. We can use definition flag in portfile to choose it.

In order to choose, one way is using vcpkg feature (`vcpkg install mongo-cxx-driver[boost]`), this is current implementation on master branch. But for unknown reason feature for C++17 was not included, so I made a PR ( https://github.com/microsoft/vcpkg/pull/22781 ) to add one more feature.

The problem is the current policy no longer allow using features to choose polyfills ( https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-use-features-to-control-alternatives-in-published-interfaces ), so my PR was not accepted.

So here is the other way I thought,
For anyone who want to specify C++ language standard, he can create a custom triplet with
```
set(VCPKG_CMAKE_CONFIGURE_OPTIONS "-DCMAKE_CXX_STANDARD=17")
```

Then in library portfile
```
if("-DCMAKE_CXX_STANDARD=17" IN_LIST VCPKG_CMAKE_CONFIGURE_OPTIONS)
    set(BSONCXX_POLY STD)
else()
    # fallback default
    set(BSONCXX_POLY MNMLSTC)
endif()
```

I am not sure if it is better way? Will it accept into master branch? (So I don't need to write overlay every time)

**Comment by Neumann-A** [comment]
> Will it accept into master branch? 

Simple answer: No. Portfiles should not assume something is injected into `VCPKG_CMAKE_CONFIGURE_OPTIONS`. If you want to inject something use the toolchain to do so.

**Reply by stalkyr** [reply]
I guessed so.

Is there any plan to specify C++ language standard in formal way? I think it will be very helpful.

**Reply by Neumann-A** [reply]
Set it in the toolchain and add the compiler flag for it.

**Comment by JackBoosY** [comment]
We accept setting the c/cxx standard as a feature, but enabling this feature must ensure that this setting is passed downstream(user's custom project) and does not cause changes to the c/cxx standard on ports that depend on it.

**Comment by dg0yt** [comment]
I'm about to update mongo-cxx-driver to v3.7.0. AFAIU the current feature design in this port is broken. But what could (be made) work with the update, given upstream's automatic choice:

Create a custom triplet with:
  ~~~cmake
  if(PORT STREQUAL "mongo-cxx-driver")
      list(APPEND VCPKG_CMAKE_CONFIGURE_OPTIONS "-DCMAKE_CXX_STANDARD=17")
  endif()
  ~~~

(For this port, other `BSONCXX_POLY_USE_<variant>` alternatives might be selected in a similar way, coupling the ABI to the triplet (right) instead of to features (wrong).

**Reply by dg0yt** [reply]
Proposed: #27565. See the comment in the portfile.
